src/opm/output/eclipse/AggregateWellData.cpp

```diff
@@ -0,0 +1,773 @@
+
+    namespace XWell {
+        std::size_t entriesPerWell(const std::vector<int>& inteHead)
+        {
+            // INTEHEAD(27) = NXWELZ
+            assert ((inteHead[27 - 1] > 123) &&
+                    "XWEL must allocate at least 124 elements per well");
+
+            return inteHead[27 - 1];
+        }
+
+        Opm::RestartIO::Helpers::WindowedArray<double>
+        allocate(const std::vector<int>& inteHead)
+        {
+            using WV = Opm::RestartIO::Helpers::WindowedArray<double>;
+
+            return WV {
+                WV::NumWindows{ numWells(inteHead) },
+                WV::WindowSize{ entriesPerWell(inteHead) }
+            };
+        }
+
+        template <class XWellArray>
+        void staticContrib(const ::Opm::Well&     well,
+                           const Opm::UnitSystem& units,
+                           const std::size_t      sim_step,
+                           XWellArray&            xWell)
+        {
+            using M = ::Opm::UnitSystem::measure;
+
+            // xWell[41] = BHP target.
+
+            xWell[41] = well.isInjector(sim_step)
+                ? well.getInjectionProperties (sim_step).BHPLimit
+                : well.getProductionProperties(sim_step).BHPLimit;
+
+            xWell[41] = units.from_si(M::pressure, xWell[41]);
+        }
+
+        template <class XWellArray>
+        void assignProducer(const std::string&         well,
+                            const ::Opm::SummaryState& smry,
+                            XWellArray&                xWell)
+        {
+            auto get = [&smry, &well](const std::string& vector)
+            {
+                return smry.get(vector + ':' + well);
+            };
+
+            xWell[0] = get("WOPR");
+            xWell[1] = get("WWPR");
+            xWell[2] = get("WGPR");
+            xWell[3] = xWell[0] + xWell[1]; // LPR
+            xWell[4] = get("WVPR");
+
+            xWell[6] = get("WBHP");
+            xWell[7] = get("WWCT");
+            xWell[8] = get("WGOR");
+
+            xWell[18] = get("WOPT");
+            xWell[19] = get("WWPT");
+            xWell[20] = get("WGPT");
+            xWell[21] = get("WVPT");
+
+            xWell[36] = xWell[1]; // Copy of WWPR
+            xWell[37] = xWell[2]; // Copy of WGPR
+        }
+
+        template <class XWellArray>
+        void assignWaterInjector(const std::string&         well,
+                                 const ::Opm::SummaryState& smry,
+                                 XWellArray&                xWell)
+        {
+            auto get = [&smry, &well](const std::string& vector)
+            {
+                return smry.get(vector + ':' + well);
+            };
+
+            // Rates reported as negative, cumulative totals as positive.
+            xWell[1] = -get("WWIR");
+            xWell[3] = xWell[1];  // Copy of WWIR
+
+            xWell[6] = get("WBHP");
+
+            xWell[23] = get("WWIT");
+
+            xWell[36] = xWell[1];  // Copy of WWIR
```

Comment:
Should there be a Bw at position 33, analogue to Bg at position 34?

Reply:
> Should there be a Bw at position 33, analogue to Bg at position 34?

It's likely, but we presently don't have solid evidence one way or the other.